internal/vehicle/carwings.go

```diff
@@ -57,52 +77,180 @@ func (v *CarWings) initSession() error {
 	v.session = &carwings.Session{
 		Region: v.region,
 	}
-
+	v.refreshKey = ""
 	return v.session.Connect(v.user, v.password)
 }
 
-// chargeState implements the api.Vehicle interface
-func (v *CarWings) chargeState() (float64, error) {
+// carwingsVehicleStatus holds the relevant data extracted from carwings API
+// on vehicle status request
+type carwingsVehicleStatus struct {
+	VehicleStatus struct {
+		SOC            float64
+		Range          int64
+		ChargingStatus carwings.ChargingStatus
+		PlugStatus     carwings.PluginState
+		ClimateStatus  struct {
+			active      bool
+			targetTemp  float64
+			outsideTemp float64
+		}
+		LastRefresh time.Time
+	}
+}
+
+func (v *CarWings) status() (res carwingsVehicleStatus, err error) {
+	// follow up requested refresh
+	if v.refreshKey != "" {
+		return v.refreshResult()
+	}
+
+	// otherwise start normal workflow
 	if v.session == nil {
 		if err := v.initSession(); err != nil {
-			return 0, api.ErrNotAvailable
+			return res, api.ErrNotAvailable
+		}
+	}
+
+	v.refreshKey, err = v.session.UpdateStatus()
+
+	if err == nil {
+		v.refreshTime = time.Now()
+	}
+	res, err = v.refreshResult()
+	if err == nil {
+		var lastUpdate time.Time
+		lastUpdate = res.VehicleStatus.LastRefresh
+
+		if elapsed := time.Since(lastUpdate); elapsed > carwingsStatusExpiry {
+			v.log.DEBUG.Printf("vehicle status is outdated (age %v > %v), requesting refresh", elapsed, carwingsStatusExpiry)
+
+			if err = v.refreshRequest(); err == nil {
+				err = api.ErrMustRetry
+			}
+		}
+	}
+
+	return res, err
+}
+
+// refreshResult triggers an update if not already in progress, otherwise gets result
+func (v *CarWings) refreshResult() (res carwingsVehicleStatus, err error) {
+
+	finished, err := v.session.CheckUpdate(v.refreshKey)
+	// update successful and completed
+	if err == nil && finished {
+		v.refreshKey = ""
+		var bs carwings.BatteryStatus
+		bs, err = v.session.BatteryStatus()
```

Comment:
why do you need to extract all of these up-front? when the refresh is valid, they are picked up by the respective call from the session anyway? looks like duplicated logic to me.

Reply:
You're right, Thanks. I am removing the duplication and keep the data in carwings.
I will test the update tomorrow during daylight and update the PR afterwards.